### PR TITLE
Repl window output perf improvements

### PR DESCRIPTION
**Pull request by mjbvz**
Currently, printing large structures to the repl window can take a very long time (#702). I believe the primary cause is we end up appending output almost one character at a time, repeating the same operations over and over again inside of a loop.

This change updates ReplWindow to take a set of outputs to be appended instead of a single output. This allows moving some of the computations out of the inner loops and reusing previously computed values.

Improves #702 . Still not great performance but it is noticeably better.

### CODE REVIEW

**Comment by mjbvz** [comment]
Confirmed that multicolor printing still seems to work:

image

File: Common/Product/ReplWindow/Repl/OutputBuffer.cs [hunk 1]
```diff
@@ -298,7 +294,7 @@ struct OutputEntryProperties {
             }
         }
 
-        enum OutputEntryKind {
+        public enum OutputEntryKind {
```
Comment (mousetraps) [inline Common/Product/ReplWindow/Repl/OutputBuffer.cs]: internal

also check usages and see if it makes sense to move into ReplWindow altogether

File: Common/Product/ReplWindow/Repl/ReplWindow.cs [hunk 2]
```diff
@@ -2110,14 +2110,18 @@ private sealed class CommandFilter : IOleCommandTarget {
         /// <summary>
         /// Appends text to the output buffer and updates projection buffer to include it.
         /// </summary>
-        internal void AppendOutput(ConsoleColor color, string text, bool lastOutput) {
+        internal void AppendOutput(OutputBuffer.OutputEntry[] entries) {
+            if (!entries.Any())
+                return;
```
Comment (mousetraps) [inline Common/Product/ReplWindow/Repl/ReplWindow.cs]: add braces

File: Common/Product/ReplWindow/Repl/ReplWindow.cs [hunk 3]
```diff
@@ -2110,14 +2110,18 @@ private sealed class CommandFilter : IOleCommandTarget {
...
+
+            var spans = new List<KeyValuePair<Span, OutputBuffer.OutputEntryProperties>>(entries.Length);
```
Comment (mousetraps) [inline Common/Product/ReplWindow/Repl/ReplWindow.cs]: move this closer to the actual usage (also check in other places). Also see if there's better variable names or something that will make the behavior more easily understood

File: Common/Product/ReplWindow/Repl/ReplWindow.cs [hunk 4]
```diff
@@ -2126,47 +2130,58 @@ private sealed class CommandFilter : IOleCommandTarget {
...
+                    startPosition += text.Length;
+                }
+                if (!_readingStdIn && !EndsWithLineBreak(text)) {
                     var lineBreak = GetLineBreak();
                     edit.Insert(oldBufferLength, lineBreak);
-                    newOutputLength += lineBreak.Length;
                     _addedLineBreakOnLastOutput = true;
-                } else {
-                    _addedLineBreakOnLastOutput = false;
+                    // Adust last span.
```
Comment (mousetraps) [inline Common/Product/ReplWindow/Repl/ReplWindow.cs]: nit: this comment doesn't add much info - either make it more specific or remove

**Comment by mousetraps** [comment]
minor comments and :+1: after that!

File: Common/Product/ReplWindow/Repl/ReplWindow.cs (hunk 3 above)
Comment (mjbvz) [inline Common/Product/ReplWindow/Repl/ReplWindow.cs]: This has to be outside of the using block here but I'll move it down a few lines.

**Pull request merged**